src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java

```diff
@@ -0,0 +1,708 @@
+                throw new HttpStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "Program is not properly configured for experiment import");
+            }
+            expSeqValue = dsl.nextval(expUnitSequenceName.toLowerCase()).toString();
+        }
+
+        for (BrAPIImport row : importRows) {
+            ExperimentObservation importRow = (ExperimentObservation) row;
+
+            PendingImportObject<BrAPITrial> trialPIO = createTrialPIO(program, commit, importRow);
+            this.trialByNameNoScope.put(importRow.getExpTitle(), trialPIO);
+
+            PendingImportObject<BrAPILocation> locationPIO = createLocationPIO(importRow);
+            this.locationByName.put(importRow.getEnvLocation(), locationPIO);
+
+            PendingImportObject<BrAPIStudy> studyPIO = createStudyPIO(program, commit, expSeqValue, importRow);
+            this.studyByNameNoScope.put(importRow.getEnv(), studyPIO);
+
+            PendingImportObject<BrAPIObservationUnit> obsUnitPIO = createObsUnitPIO(program, commit, obsUnitNextVal, importRow);
+            String key = createObservationUnitKey(importRow);
+            this.observationUnitByNameNoScope.put(key, obsUnitPIO);
+        }
+    }
+
+    private String createObservationUnitKey(ExperimentObservation importRow) {
+        String key = importRow.getEnv() + importRow.getExpUnitId();
+        return key;
+    }
+
+    private ValidationErrors validateFields(List<BrAPIImport> importRows, ValidationErrors validationErrors) {
+        HashSet<String> uniqueStudyAndObsUnit = new HashSet<>();
+        for (int i = 0; i < importRows.size(); i++) {
+            ExperimentObservation importRow = (ExperimentObservation) importRows.get(i);
+            validateConditionallyRequired(validationErrors, i, importRow);
+            validateUniqueObsUnits(validationErrors, uniqueStudyAndObsUnit, i, importRow);
+        }
+        return validationErrors;
+    }
+
+    /**
+     * Validate that the the observation unit is unique within a study.
+     * <br>
+     * SIDE EFFECTS:  validationErrors and uniqueStudyAndObsUnit can be modified.
+     * @param validationErrors can be modified as a side effect.
+     * @param uniqueStudyAndObsUnit can be modified as a side effect.
+     * @param i counter that is always two less the file row being validated
+     * @param importRow the data row being validated
+     */
+    private void validateUniqueObsUnits(
+            ValidationErrors validationErrors,
+            HashSet<String> uniqueStudyAndObsUnit,
+            int i,
+            ExperimentObservation importRow) {
+        String envIdPlusStudyId = createObservationUnitKey( importRow );
+        if( uniqueStudyAndObsUnit.contains( envIdPlusStudyId )){
+            String errorMessage = String.format("The ID (%s) is not unique within the environment(%s)", importRow.getExpUnitId(), importRow.getEnv());
```

Comment:
Instead of "The ID (%s)" it might be more useful to have the actually column name.

Reply:
The actually column name (Exp Unit ID) will be included in the message to the user via the next line

> this.addRowError("Exp Unit ID", errorMessage, validationErrors, i);